addons/account/models/account_account.py

```diff
@@ -999,37 +997,80 @@ def write(self, vals):
+              This can be checked by checking that the code is set for each company in account.company_ids.root_id.
+
+            - for each `company` in `account.company_ids`, there must not be any `other_account` that is:
+                - accessible from a same company as `account`; and
+                - has the same code as `account` in a company from which both are accessible.
+
+              Noting that companies can access all the accounts of their parents, we can check this by ensuring that there
+              is no `other_account` that belongs to a parent of `account.company_ids` or a child of `account.company_ids`,
+              and that have the same code in the root company.
+
+            For example, if account.company_ids == res.company(1), then, we check:
+            - account.with_company(1).code != False
+            - self.with_company(res.company(1).root_id).search([
+                ('company_ids', 'in' <parents and children of res.company(1)>),
+                ('code', '=', account.with_company(1).code),
+              ]) must be empty.
+
+            :param check_all_companies: If False, we will perform the check only for the currently active
+                                        company. Pass this param if you know that you only modified the code
+                                        in the currently active company and that the company_ids did not change.
+        """
+        accounts_to_check = self.sudo() if check_all_companies else self.sudo().filtered(lambda a: self.env.company in a.company_ids)
+
+        # Check 1: Check that the code is set.
+        for account in accounts_to_check:
+            for company in account.company_ids.root_id:
                 if not account.with_company(company).code:
                     raise ValidationError(_("The code must be set for every company to which this account belongs."))
-        accounts_to_check = accounts.filtered(lambda a: a.code and self.env.company in a.company_ids)
-        accounts_by_code = accounts_to_check.grouped('code')
-        duplicate_codes = None
-        if len(accounts_by_code) < len(accounts_to_check):
-            duplicate_codes = [code for code, accounts in accounts_by_code.items() if len(accounts) > 1]
-        # search for duplicates of self in database
-        elif duplicates := self.sudo().search_fetch(
-            [
-                ('code', 'in', list(accounts_by_code)),
-                ('id', 'not in', self.ids),
-            ],
-            ['code'],
-        ):
-            duplicate_codes = duplicates.mapped('code')
-        if duplicate_codes:
-            raise ValidationError(
-                _("Account codes must be unique. You can't create accounts with these duplicate codes: %s", ", ".join(duplicate_codes))
-            )
+
+        # Do a grouping by companies in `company_ids`.
+        accounts_to_check_by_company = defaultdict(self.env['account.account'].browse)
```

Comment:
This will run in 0(n²) because recordsets are immutable

Reply:
Ohhh thanks. I've rewritten it by constructing a list of ids. I don't know if there's a better way to do it?